cli/main.rs

```diff
@@ -178,114 +181,273 @@ fn print_cache_info(state: &GlobalState, json: bool) -> Result<(), ErrBox> {
   }
 }
 
-struct FileInfoOutput<'a> {
-  local: &'a str,
-  file_type: &'a str,
-  compiled: Option<String>,
-  map: Option<String>,
-  deps: Option<Deps>,
+/// A dependency tree of the basic module information.
+///
+/// Constructed from a `ModuleGraph` and `ModuleSpecifier` that
+/// acts as the root of the tree.
+struct FileInfoDepTree {
+  name: String,
+  size: usize,
+  total_size: Option<usize>,
+  deps: Vec<FileInfoDepTree>,
 }
 
-// TODO(bartlomieju): this function de facto repeats
-// whole compilation stack. Can this be done better somehow?
+impl serde::Serialize for FileInfoDepTree {
+  fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+  where
+    S: serde::Serializer,
+  {
+    let mut state = serializer.serialize_struct("FileInfoDepTree", 4)?;
+
+    state.serialize_field("name", &self.name)?;
+    if let Some(total_size) = self.total_size {
+      state.serialize_field("size", &self.size)?;
+      state.serialize_field("total_size", &total_size)?;
+    }
+    state.serialize_field("deps", &self.deps)?;
+    state.end()
+  }
+}
+
+impl FileInfoDepTree {
+  /// Create a `FileInfoDepTree` tree from a `ModuleGraph` and the root `ModuleSpecifier`.
+  pub fn new(
+    module_graph: &ModuleGraph,
+    root_specifier: &ModuleSpecifier,
+  ) -> Self {
+    let mut seen = HashSet::new();
+    let mut total_sizes = HashMap::new();
+
+    Self::cstr_helper(&mut seen, &mut total_sizes, module_graph, root_specifier)
+  }
+
+  fn cstr_helper(
+    seen: &mut HashSet<String>,
+    total_sizes: &mut HashMap<String, usize>,
+    graph: &ModuleGraph,
+    root: &ModuleSpecifier,
+  ) -> Self {
+    let name = root.to_string();
+    let never_seen = seen.insert(name.clone());
+    let file = graph.get(&name).unwrap();
+    let size = file.size();
+    let deps = if never_seen {
+      file
+        .imports
+        .iter()
+        .map(|import| import.resolved_specifier.clone())
+        .map(|spec| Self::cstr_helper(seen, total_sizes, graph, &spec))
+        .collect::<Vec<_>>()
+    } else {
+      vec![]
+    };
+
+    let total_size = if never_seen {
+      if let Some(total_size) = total_sizes.get(&name) {
+        Some(total_size.to_owned())
+      } else {
```

Comment:
That's a lot of new code, it makes sense to create a new module called `info.rs` and move all of that logic there

Reply:
Sounds good. Thanks for the review.